### ISSUE TITLE
LiteDB ASP.NET CORE 2 configuration

### ISSUE THREAD

**Issue by mhnegrao** [litedb-org/LiteDB#1035]
Hello Mauricio, I am implementing a small application with asp.net.cor 2.0, to test the new technology "Blazor", which is very cool and efficient and I am not succeeding.
Are there any additional settings to be used in NET.CORE 2?

tks!
Marcelo Henrique

**Comment by mbdavid** [comment]
Hi @mhnegrao, I heard about this technology but I didn't test yet. Current v4 support .net standard 2 (and are compiled on this version).

I current using v5 (dev version) in a new small app using asp.net core 2.0 (with .net core 2) with no problem. Just copy .dll file into my bin folder and works.

If you can use LiteDB in Blazor, please let me know so I can set an example for the community.

**Comment by mhnegrao** [comment]
Hi ... I'm going to speak in Portuguese ... to flow better ... rs rs
vamos lá: qdo tento acessar o arquivo do banco de dados, parece que a aplicação não consegue criar o arquivo .db. ao a brir as ferramentas do desenvolvedor vejo este erro no console: WASM: [System.Runtime.Serialization.SerializationException] Invalid JSON string, pois não há dados e nem arquivo pra retornar o json. Este é o meu modelo de dados:

public class OrdemServico
    {
        public OrdemServico()
        {
            GerarDados();
        }
        public int Id { get; set; }
        public DateTime DataOS { get; set; }
        public string Descricao { get; set; }
        public decimal Valor { get; set; }
        public bool Executada { get; set; }
        public DateTime ExecutadoEm { get; set; }
        public List<ServicoExecutado> ServicosExewcutados { get; set; } = new List<ServicoExecutado>();

        public void GerarDados()
        {
            this.Id = 1;
            this.Descricao = "Teste";
            this.DataOS = DateTime.Now;
            this.Executada = true;
            this.ExecutadoEm = DateTime.Now;
            this.Valor = 1000;
        }

Será que esqueci algo?

**Comment by mbdavid** [comment]
cara, não vi no teu código a chamado no LiteDB... mas cuida em relação aonde tu quer gerar o arquivo, pois teu usuario precisa ter permissão nesse diretorio.

**Comment by mbdavid** [comment]
de cara vejo 2 problemas:

1) tu tá gravando o banco no wwwroot. Tens permissão neste diretorio? geralmente não se tem pq essa é a area publica.... nao deveria ter acesso externo.

2) se tu abre/fecha o banco de dados não pode quere retornar um IEnumerable.... pois antes de chamar o primeiro registro tu ja fechou o banco

        public IEnumerable<TEntity> GetAll()
        {
            using (var db = new LiteDatabase(database))
            {
                var regs =
db.GetCollection<TEntity>(typeof(TEntity).FullName).FindAll();
                return regs;
            }
        }

Se tu quer retornar todos os dados e depois fechar o banco, precisa usar um "ToArray()" ou "ToList()" depois do FindAll()

Tu chegou a tentar abrir o banco com a ferramenta shell pra ver se tem algum registro la?

**Comment by mhnegrao** [comment]
Sim... abri o banco e não tinha nada...nenhuma coleção... só banco vazio.
Agora pedindo uma ajuda: Em Blazor existem 3 projetos criados inicialmente no modo hosted:

proj.client
proj.server
proj.shared

e  efetuando os testes verifiquei que não estou conseguindo gerar o arquivo corretamente.
No caso desta aplicação, wwwroot, onde está o banco está no projeto proj.server que é o que faz as requisições no server (esta pasta eu criei manualmente pra testar). Exite a pasta wwwroot em proj.client que é o início da aplicação cliente e é criada por padrão.
Vou efetuar as alterações que vc me indicou. Agora vai a pergunta: qual a melhor maneira para eu criar o arquivo no projeto proj.server, que é onde o repositório de dados está?
O asp.net core tem algumas particularidades que talvez eu não esteja observando, já que o banco irá ficar na web.

**Comment by mhnegrao** [comment]
Se eu conseguir solucionar estes probleminhas, eu passo para vc em detalhes a configuração apropriada para se usar com Blazor

**Comment by AngelMunoz** [comment]
Hello there, I don't speak Portuguese but from what I pick up from the bits of the conversation (I speak Spanish)
LiteDB is unlikely to work on WASM based projects (blazor, uno wasm) since LiteDB needs to write to disk (which is unavailable in websites). LiteDB, however should be able to work just fine on a asp.net core project (the server side) as long as the server application has access to the directory the database is located on

**Comment by JKamsker** [comment]
A few things in the code above explain the empty file and the JSON error. None of them are LiteDB settings.

1. **The collection name is invalid.** `typeof(TEntity).FullName` is `AppDatabase.Shared.Models.Cliente`, and collection names can't contain dots. `Insert` throws `LiteException: Invalid collection name 'AppDatabase.Shared.Models.Cliente': Use only [a-Z$_]` (4.x throws `Invalid format: ...`). The file gets created, but nothing is ever saved. Use `typeof(TEntity).Name` or a fixed name.
2. **Check the routes.** The controller has `[Route("api/cliente")]`, so the action route `[Route("api/cliente/Index")]` is appended to it and becomes `api/cliente/api/cliente/Index`. A request to `/api/cliente/index` is matched by `Get(int id)` instead and returns `204 No Content` with an empty body, so `GerarDados` never runs. That empty body would explain the "Invalid JSON string" on the client. Use `[HttpGet("Index")]` on the action.
3. **Don't return a lazy query from a `using` block.** `FindAll()` is deferred, so `GetAll()` hands back a query whose database is already disposed. On 5.0.21 and the 6.x prerelease, enumerating it throws `ObjectDisposedException`. Call `.ToList()` before returning.
4. **Keep the file out of `wwwroot`**, in a folder the app can write to, and use one shared instance instead of opening the database on every call:

```csharp
// Program.cs (server)
var dataDir = Path.Combine(builder.Environment.ContentRootPath, "App_Data");
Directory.CreateDirectory(dataDir);
builder.Services.AddSingleton<ILiteDatabase>(_ =>
    new LiteDatabase(Path.Combine(dataDir, "MeusDados.db")));

// repository: inject ILiteDatabase _db
public List<TEntity> GetAll() =>
    _db.GetCollection<TEntity>(typeof(TEntity).Name).FindAll().ToList();
```

`LiteDatabase` is thread-safe, and the DI container disposes it on shutdown. I ran your controller and repository in ASP.NET Core to confirm the first three points, and this setup with parallel requests, on 5.0.21 and the 6.x prerelease. LiteDB belongs in the server project, as you have it. Blazor WebAssembly code runs in the browser and can't reach the server's disk.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
